### PR TITLE
[WIP] Combat Framework

### PR DESCRIPTION
**Pull request by uwadmin12**
## Current Progress

This PR is for Stage 1 only

### Stage 1

- [x] Restructure AttackAction cycle (locate target -> queue -> completion)
- [x] Creature faction and hostility check
- [x] Scanning area for hostiles and add those to combatant list
- [ ] Equipment Swapping Commands
- [ ] Crack the combat animation code (match animation to current equipment)
- [x] **Combat Mode toggle**
- [ ] Combat-mode default animation
- [x] Use deactivation timer (3 second timeout for inactivity)
- [x] Use ai timer (3 second for each creature/rotate per frame)???
- [ ] Implement script functions for GetIsEnemy and SetFaction, etc.
- [ ] (Optional) Invoke k_ai_master

### Stage 2

- [ ] \[D20\] OG GUI for combat mode
- [ ] \[D20\] Damage/effect logic and death
- [ ] Feedback text, animated text (damage)
- [ ] Item Consumption Commands
- [ ] Force Power implementation
- [ ] AI Optimization: Max Damage Output
- [ ] AI Optimization: Max Buff onEnterCombat
- [ ] AI Optimization: Item Consumption for survival
- [ ] (TBD) AI Optimization: Keep track of last known position of hostile on deactivation timer start (for realism)

### Stage 3

- [ ] Integrate visual effect for all types of attack
- [ ] \[Overhaul\] Subclass for control overhaul and real-time combat

Demo 1 (10/19/2020): https://youtu.be/beLJEoUAQmg

### CODE REVIEW

**Comment by seedhartha** [comment]
Great job! I think the overall idea is there. Now we need to refactor it some 😅 Hopefully I don't come as too nitpicky.

**Comment by uwadmin12** [comment]
Oh yeah, it's pretty messy at the moment ... I dumped everything into game.cpp to see what dependencies it will need.
Any advice/directions would be much appreciated!

**Comment by seedhartha** [comment]
Please follow the coding style of the project. Something that catched my eye:

- Class member variables should begin with an underscore.
- Function and variable names should be in camelCase.

**Comment by seedhartha** [comment]
Diagrams of the combat system after refactoring, as discussed in Discord.

[](https://images2.imgbox.com/e1/39/cnJL5fFG_o.jpg)
[](https://images2.imgbox.com/53/e1/u8AuftH9_o.jpg)

File: src/game/combat.h [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+#include "types.h"
+#include "object/creature.h"
+
+namespace reone {
+
+namespace game {
+
+class Area;
+
+class CombatManager {
```
Comment (seedhartha) [inline src/game/combat.h:32]: Just `Combat` would be a better name, I believe. Manager classes are considered an anti-pattern, as it is too vague of a term.

File: src/game/combat.cpp [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+// }
+
+CombatManager::CombatManager(Area *area) : _area(area) {
+    if (!area) {
+        throw invalid_argument("Area must not be null");
+    }
+}
+
+void CombatManager::update(float dt)
+{
```
Comment (seedhartha) [inline src/game/combat.cpp:54]: Opening brackets should be on the same line with function declarations.

File: src/game/combat.cpp [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+            combatStateMachine(cbt);
+        }
+        animationControl();
+    }
+    else {
+        debug("no pc yet ...");
+    }
+}
+
+void CombatManager::queueInstructions(std::shared_ptr<Creature> &combatant)
```
Comment (seedhartha) [inline src/game/combat.cpp:72]: Here and below, `std::` can be omitted.

File: src/game/object/creature.h [hunk 4]
```diff
@@ -90,6 +98,31 @@ class Creature : public SpatialObject {
...
+    /* combat animation interruption */
+    bool isInterrupted() {
+        return cbt_s != CombatState::Idle;
+    }
+
+    void setStaggered() {
+        staggered = true;
+    }
+
+    bool isStaggered() {
```
Comment (seedhartha) [inline src/game/object/creature.h:115]: Getter functions (`get*`, `is*`, `has*`) should not have side effects. I think, toggle logic should be moved to the calling code.

File: src/game/object/creature.h [hunk 5]
```diff
@@ -90,6 +98,31 @@ class Creature : public SpatialObject {
 
     // END Pathfinding
 
+    CombatState cbt_s = CombatState::Idle;
+
+    bool melee_command = false;
```
Comment (seedhartha) [inline src/game/object/creature.h:103]: Member variables should not have public access, as it breaks encapsulation.

**Review by seedhartha — CHANGES_REQUESTED** [review]

File: src/game/action/action.h [hunk 6]
```diff
@@ -33,7 +33,9 @@ enum class ActionType {
     StartConversation = 0x1001,
     PauseConversation = 0x1002,
     ResumeConversation = 0x1003,
-    MoveToObject = 0x1004
+    MoveToObject = 0x1004,
+
+    AttackAction
```
Comment (seedhartha) [inline src/game/action/action.h:38]: To match _nwscript.nss_ we should name it `AttackObject` and give it a value of 3. This action could contain a type of attack in the future.

File: src/game/actionexecutor.cpp [hunk 7]
```diff
@@ -66,6 +66,9 @@ void ActionExecutor::executeActions(Object &object, float dt) {
         case ActionType::StartConversation:
             executeStartConversation(static_cast<Creature &>(object), *dynamic_cast<StartConversationAction *>(action), dt);
             break;
+        case ActionType::AttackAction:
+            executeAttack(static_cast<Creature&>(object), *dynamic_cast<AttackAction*>(action), dt);
```
Comment (seedhartha) [inline src/game/actionexecutor.cpp:70]: Let's use the following code style (whitespace after type name, not before variable name):

```cpp
Creature &creature;
AttackAction *action;
```

File: src/game/combat.h [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+    * 
+    * [Future Roles]:
+    * 1. Evaluate the possibility and extent of damage infliction
+    * 2. Evaluate the possibility and duration of stagger
+    * 3. Share the responsibility of setting timers
+    */
+    void animationControl();
+
+private:
+    Area*                                                       _area;
```
Comment (seedhartha) [inline src/game/combat.h:67]: nit: Let's not pad variable names. I find it time consuming and not really beneficial to readability.

**Review by seedhartha — CHANGES_REQUESTED** [review]

File: src/game/action/attack.h [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+        ObjectAction(ActionType::AttackAction, object), _timeout(timeout), _inPursuit(pursue), _distance(distance) { }
+
+    std::shared_ptr<Creature> target() { return std::static_pointer_cast<Creature>(_object); } // static_pointer_cast(nullptr)?
+
+    void setAttack() { _inPursuit = false;  }
+
+    /* not in pursuit means attack signal is high, be it ranged or melee */
+    bool isInPursuit() { return _object != nullptr && _inPursuit; }
+
+    float& getTimeout() { return _timeout; }
```
Comment (seedhartha) [inline src/game/action/attack.h:39]: This breaks encapsulation and is a little counter-intuitive 🙂 In OOP, objects should receive commands and act on their (hidden) internal state. How about this?

```cpp
void advance(float dt) {
    _timeout = glm::max(0.0f, _timeout - dt);
}

bool isTimedOut() const {
    return _timeout == 0.0f;
}
```

File: src/game/object/creature.h [hunk 10]
```diff
@@ -43,13 +52,13 @@ class Creature : public SpatialObject {
...
+        glm::vec3 destination{ 0.0f };
         std::vector<glm::vec3> points;
-        uint32_t timeFound { 0 };
-        int pointIdx { 0 };
+        uint32_t timeFound{ 0 };
+        int pointIdx{ 0 };
     };
 
-    Creature(uint32_t id, ObjectFactory *objectFactory, scene::SceneGraph *sceneGraph);
+    Creature(uint32_t id, ObjectFactory* objectFactory, scene::SceneGraph* sceneGraph);
```
Comment (uwadmin12) [inline src/game/object/creature.h:61]: Hmm I don't know how those happened 😕

File: src/game/effect.h [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+        : _type(type), _timeout(duration) {}
+
+    virtual ~Effect() = default;
+
+    void update(float dt) {
+        _timeout -= dt;
+    }
+
+    /* Should keep a global SDL_GetTick() per frame */
+    bool isValid(uint32_t currentTick) const { return _timeout > 0 ; }
```
Comment (uwadmin12) [inline src/game/effect.h:45]: Remove `uint32_t currentTick`

File: src/game/object/creature.h [hunk 12]
```diff
@@ -79,25 +88,43 @@ class Creature : public SpatialObject {
...
+    CombatState getCombatState() { return _cbtState; }
+    void setCombatState(CombatState state) { _cbtState = state;  }
+
+    // const std::deque<std::unique_ptr<Effect>> &getActiveEffects() { return _activeEffects; }
+
+    void applyEffect(std::unique_ptr<Effect>&& eff) {
+        _activeEffects.push_back(std::move(eff)); 
+    }
+
+    std::shared_ptr<Creature> currentEnemy();
```
Comment (uwadmin12) [inline src/game/object/creature.h:124]: Remove this function.

File: src/game/effect.h [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+#include <memory>
+
+namespace reone {
+
+namespace game {
+
+class Creature;
+
+enum class EffectType {
+    EFFECT_TYPE_INVALIDEFFECT = 0,
```
Comment (seedhartha) [inline src/game/effect.h:29]: Could be just `Invalid`.

File: src/game/effect.h [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+};
+
+/* Calculate damage based on equipment and effects */
+class DamageEffect : public Effect {
+public:
+    DamageEffect(const std::shared_ptr<Creature>& applicator)
+        : Effect(EffectType::EFFECT_TYPE_INVALIDEFFECT, 0), _applicator(applicator) {}
+
+    /* for feedback text, mostly */
+    std::shared_ptr<Creature>& getApplicator() { return _applicator; }
```
Comment (seedhartha) [inline src/game/effect.h:64]: `Inflictor` is more suitable than `applicator`, I think.

File: src/game/combat.h [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+    constexpr bool operator()(
+        std::pair<uint32_t, T> const& a,
+        std::pair<uint32_t, T> const& b)
+        const noexcept {
+        return a.first > b.first;
+    } // min heap!
+};
+
+template <typename T>
+struct TimerQueue {
```
Comment (seedhartha) [inline src/game/combat.h:48]: Given its generic nature, shouldn't this class be in libcommon?

File: src/game/combat.cpp [hunk 16]
```diff
@@ -0,0 +1,234 @@
...
+    while (!_bashQueue.empty()) {
+        auto& pr = _bashQueue.front();
+        duel(pr.first, pr.second);
+        _bashQueue.pop_front();
+    }
+}
+
+shared_ptr<Creature> Combat::findNearestHostile(shared_ptr<Creature> &combatant) {
+    shared_ptr<Creature> closest_target = nullptr;
+    float min_dist = 1000000;
```
Comment (seedhartha) [inline src/game/combat.cpp:215]: `FLT_MAX`

File: src/game/action/attack.h [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+    void setAttack() { _inRange = true;  } // handled by actionExecutor
+
+    bool isInRange() { return _inRange; }
+
+    void advance(float dt) {
+        _timeout = glm::max(0.0f, _timeout - dt);
+    }
+
+    bool isTimedOut() const {
+        return _timeout < 1E-6;
```
Comment (seedhartha) [inline src/game/action/attack.h:43]: Why not compare with `0.0f`? Timeout's lower bound is exactly 0.0.

**Review by seedhartha — CHANGES_REQUESTED** [review]

File: src/game/action/attack.h (hunk 17 above)
Comment (uwadmin12) [reply src/game/action/attack.h:43]: I don't trust float point equality, even in this case.

File: src/game/combat.h (hunk 15 above)
Comment (uwadmin12) [reply src/game/combat.h:48]: Good idea.

File: src/game/effect.h (hunk 14 above)
Comment (seedhartha) [reply src/game/effect.h:64]: Actually, `damager`. This is from nwscript.nss, better to follow their terminology.

**Comment by seedhartha** [comment]
Merged, thanks! I did a minor refactoring job, fixed CI compilation errors and code style issues. Combat is disabled for now, because is kind of breaks the gameplay.

**Pull request merged**